ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitter.scala

```diff
@@ -387,9 +388,29 @@ private[kvutils] class TransactionCommitter(
     )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
       val transaction = transactionEntry.submission.getTransaction
       val nodes = transaction.getNodesList.asScala
-      val nodesToKeep = nodes.iterator.collect {
-        case node if node.hasCreate || node.hasExercise => node.getNodeId
-      }.toSet
+      val nodeMap: Map[String, TransactionOuterClass.Node] =
+        nodes.view.map(n => n.getNodeId -> n).toMap
+
+      @tailrec
+      def goNodesToKeep(todo: List[String], result: Set[String]): Set[String] = todo match {
+        case Nil => result
+        case head :: tail =>
+          import TransactionOuterClass.Node.NodeTypeCase
+          val node = nodeMap
+            .get(head)
+            .getOrElse(throw Err.InvalidSubmission(s"Invalid transaction node id $head"))
```

Comment:
Can this really happen because of an invalid submission? It seems a rather impossible case to me as we build this map with the key being a piece of the node; we could perhaps have clashes if things are really broken. Perhaps I'd either throw an internal error or even let Scala throw if the element is absent.

Reply:
I don’t really know what validation happens up to this point so I don’t know for sure but I expect you’re right. I’ve switched it to `InternalError`.

Reply:
Thanks, this also increases our incentive to look into these cases, as we'll tend to assume it's a bug.